### PR TITLE
Add logger to startup

### LINKED ISSUE #1864: Add Log support to StartupHook and AspNetCoreBootstrapper
Both `OpenTelemetry.AutoInstrumentation.StartupHook` and `OpenTelemetry.AutoInstrumentation.AspNetCoreBootstrapper` writes the status to `EventSource`. Additional tools like `logman`, `dotnet-trace`, `Perfview`, etc., are needed to capture these event source logs.

We have an in-built file-based logging in [`OpenTelemetry.AutoInstrumentation`](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/tree/main/src/OpenTelemetry.AutoInstrumentation/Logging) project. We should extend this feature to StartupHook and AspNetCoreBootstrapper to write the logs to file system for easier debugging.

### PR DESCRIPTION
**Pull request by dszmigielski**
## Why

We do not log to the file in `StartupHook` and `AspNetCoreBootstrapper`.
 
Fixes #1864

## What

Extract logging logic to separate project. Use file logging in the code mentioned above.

## Tests

CI/CD

## Checklist

- ~~[ ] `CHANGELOG.md` is updated.~~
- ~~[ ] Documentation is updated.~~
- ~~[ ] New features are covered by tests.~~

### CODE REVIEW

File: src/OpenTelemetry.AutoInstrumentation.Logging/Properties/AssemblyInfo.NetFramework.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+#if NETFRAMEWORK
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+[assembly: AssemblyKeyFile("keypair.snk")]
```
Comment (pellared) [inline src/OpenTelemetry.AutoInstrumentation.Logging/Properties/AssemblyInfo.NetFramework.cs:21]: Can you check if you can move move `keypair.snk` to `src` and change this like to:


```suggestion
[assembly: AssemblyKeyFile("..\\keypair.snk")]
```

The same should be then done in `src\OpenTelemetry.AutoInstrumentation.Logging\Properties\AssemblyInfo.NetFramework.cs`

Reference: https://learn.microsoft.com/en-us/dotnet/api/system.reflection.assemblykeyfileattribute.-ctor?view=net-7.0

File: src/OpenTelemetry.AutoInstrumentation.Logging/OpenTelemetry.AutoInstrumentation.Logging.csproj [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <Import Project="..\CommonExcludedAssets.props" />
+
+  <PropertyGroup>
+    <Version>0.5.1</Version>
```
Comment (pellared) [inline src/OpenTelemetry.AutoInstrumentation.Logging/OpenTelemetry.AutoInstrumentation.Logging.csproj:6]: Option 1: Update `releasing.md` with info to bump the version to.

Option 2: Maybe we can mve this property to `src/Directory.Build.props`?

File: src/OpenTelemetry.AutoInstrumentation.AspNetCoreBootstrapper/Logger.cs [hunk 3]
```diff
@@ -14,23 +14,23 @@
...
+    private static readonly ILogger Log = OtelLogging.GetLogger();
 
-    /// <summary>
-    /// Gets the column number of the document where the error occurred
-    /// </summary>
-    public int Column;
+    internal static void LogTrace(string message)
+    {
+        Log.Information(message);
+        BootstrapperEventSource.Log.Trace(message);
```
Comment (pellared) [inline src/OpenTelemetry.AutoInstrumentation.AspNetCoreBootstrapper/Logger.cs:28]: Not directly related, but I think we should make the logging and EventSource usage consistent. `AutoInstrumentationEventSource` looks to be used in a different way. Additionally, probably we could have a single base class for all of our event sources.

I have also noticed that all event sources' `Trace` methods have `Level = EventLevel.Informational` (sic!). We should clean up these levels.

This comment should be addressed in a seperate PR. If you agree, you can create an issue out of this comment.

File: src/OpenTelemetry.AutoInstrumentation.AspNetCoreBootstrapper/Logger.cs [hunk 4]
```diff
@@ -14,23 +14,23 @@
...
-    /// Gets the line number of the document where the error occurred
-    /// </summary>
-    public int Line;
+    private static readonly ILogger Log = OtelLogging.GetLogger();
 
-    /// <summary>
-    /// Gets the column number of the document where the error occurred
-    /// </summary>
-    public int Column;
+    internal static void LogTrace(string message)
```
Comment (pellared) [inline src/OpenTelemetry.AutoInstrumentation.AspNetCoreBootstrapper/Logger.cs:25]: It is not a "trace level" log, but a "Information" level log. Can we rename it to `Information` or `LogInformation` or `LogInfo`?

File: src/OpenTelemetry.AutoInstrumentation.Logging/CustomLogger.cs [hunk 5]
```diff
@@ -14,7 +14,7 @@
 // limitations under the License.
 // </copyright>
 
-using System.Runtime.CompilerServices;
+using System;
```
Comment (pellared) [inline src/OpenTelemetry.AutoInstrumentation.Logging/CustomLogger.cs:17]: I think we can rename `CustomLogger` to `Logger`. Then we could move the `OtelLogging` code here and rename `GetLogger()` to `Get()` or even `Value`.

This shoud be addressed in a separate PR.

File: src/OpenTelemetry.AutoInstrumentation.StartupHook/Logger.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+{
+    private static readonly ILogger Log = OtelLogging.GetLogger();
+
+    internal static void LogTrace(string message)
+    {
+        StartupHookEventSource.Log.Trace(message);
+
+        try
+        {
+            Log.Information(message);
```
Comment (pellared) [inline src/OpenTelemetry.AutoInstrumentation.StartupHook/Logger.cs:32]: Shouldn't it be `Log.Trace(message);`?

File: src/OpenTelemetry.AutoInstrumentation.StartupHook/StartupHook.cs [hunk 7]
```diff
@@ -75,20 +77,31 @@ public static void Initialize()
...
+    private static void InitLogger()
+    {
+        // Workaround to make startup hook load the logger assembly.
+        var executionPath = Assembly.GetExecutingAssembly().Location;
+        var dirname = Path.GetDirectoryName(executionPath);
+        if (!string.IsNullOrEmpty(dirname))
+        {
+            Assembly.LoadFrom(Path.Combine(dirname, "OpenTelemetry.AutoInstrumentation.Logging.dll"));
+        }
+    }
```
Comment (pellared) [inline src/OpenTelemetry.AutoInstrumentation.StartupHook/StartupHook.cs:103]: How about moving this code to the `Logger` static constructor?

File: src/OpenTelemetry.AutoInstrumentation.StartupHook/StartupHook.cs [hunk 8]
```diff
@@ -75,20 +77,31 @@ public static void Initialize()
...
     }
 
+    private static void InitLogger()
+    {
+        // Workaround to make startup hook load the logger assembly.
+        var executionPath = Assembly.GetExecutingAssembly().Location;
+        var dirname = Path.GetDirectoryName(executionPath);
+        if (!string.IsNullOrEmpty(dirname))
+        {
+            Assembly.LoadFrom(Path.Combine(dirname, "OpenTelemetry.AutoInstrumentation.Logging.dll"));
```
Comment (pellared) [inline src/OpenTelemetry.AutoInstrumentation.StartupHook/StartupHook.cs:101]: Shouldn't we catch exception thrown by `Assembly.LoadFrom` and use event source to have any kind of indcation that something bad happened?

File: src/OpenTelemetry.AutoInstrumentation.StartupHook/Logger.cs (hunk 6 above)
Comment (dszmigielski) [reply src/OpenTelemetry.AutoInstrumentation.StartupHook/Logger.cs:32]: Why? This is the file logger, it does not have a `Trace` method.

File: src/OpenTelemetry.AutoInstrumentation.StartupHook/StartupHook.cs (hunk 7 above)
Comment (dszmigielski) [reply src/OpenTelemetry.AutoInstrumentation.StartupHook/StartupHook.cs:103]: Tried that and it didn't work for some reason :|

File: src/OpenTelemetry.AutoInstrumentation.StartupHook/Logger.cs (hunk 6 above)
Comment (pellared) [reply src/OpenTelemetry.AutoInstrumentation.StartupHook/Logger.cs:32]: I see. This can be addressed together with https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/1983/files/19815362edfaf339754bbd7eb8c8c15288e0e50e#r1072147421

File: src/OpenTelemetry.AutoInstrumentation.Logging/OpenTelemetry.AutoInstrumentation.Logging.csproj (hunk 2 above)
Comment (nrcventura) [reply src/OpenTelemetry.AutoInstrumentation.Logging/OpenTelemetry.AutoInstrumentation.Logging.csproj:6]: It might be simpler to use option 1 for now, and perhaps try out option 2 separately.

File: src/OpenTelemetry.AutoInstrumentation.Logging/ILogger.cs [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+// limitations under the License.
+// </copyright>
+
+using System;
+
+namespace OpenTelemetry.AutoInstrumentation.Logging;
+
+#pragma warning disable CS1591
+[System.Diagnostics.CodeAnalysis.SuppressMessage("StyleCop.CSharp.DocumentationRules", "SA1600:Elements should be documented", Justification = "Not")]
+public interface ILogger
```
Comment (rajkumar-rangaraj) [inline src/OpenTelemetry.AutoInstrumentation.Logging/ILogger.cs:23]: Shall we use a different name here?
`ILogger` is a dotnet reserved name and we could avoid naming conflict.

File: src/OpenTelemetry.AutoInstrumentation.StartupHook/StartupHook.cs [hunk 10]
```diff
@@ -75,20 +77,38 @@ public static void Initialize()
...
+    private static void InitLogger()
+    {
+        // Workaround to make startup hook load the logger assembly.
+        var executionPath = Assembly.GetExecutingAssembly().Location;
+        var dirname = Path.GetDirectoryName(executionPath);
+        if (!string.IsNullOrEmpty(dirname))
+        {
+            try
+            {
+                Assembly.LoadFrom(Path.Combine(dirname, "OpenTelemetry.AutoInstrumentation.Logging.dll"));
```
Comment (rajkumar-rangaraj) [inline src/OpenTelemetry.AutoInstrumentation.StartupHook/StartupHook.cs:103]: `OpenTelemetry.AutoInstrumentation.StartupHook.csproj` has project reference to `OpenTelemetry.AutoInstrumentation.Logging.csproj`, hence `OpenTelemetry.AutoInstrumentation.Logging.dll` should get loaded. Is there a reason to load this dynamically?

Comment (dszmigielski) [reply src/OpenTelemetry.AutoInstrumentation.StartupHook/StartupHook.cs:103]: The reason is it does not load it for some reason. My guess is that StartupHook is loaded as `DOTNET_STARTUP_HOOKS` and it just loads single assembly.

File: src/OpenTelemetry.AutoInstrumentation.Logging/ILogger.cs (hunk 9 above)
Comment (dszmigielski) [reply src/OpenTelemetry.AutoInstrumentation.Logging/ILogger.cs:23]: I'll rename to IOtelLogger.

File: src/OpenTelemetry.AutoInstrumentation.Logging/OpenTelemetry.AutoInstrumentation.Logging.csproj (hunk 2 above)
Comment (dszmigielski) [reply src/OpenTelemetry.AutoInstrumentation.Logging/OpenTelemetry.AutoInstrumentation.Logging.csproj:6]: I have updated `releasing.md` for now.

File: src/OpenTelemetry.AutoInstrumentation.StartupHook/StartupHook.cs (hunk 10 above)
Comment (pellared) [reply src/OpenTelemetry.AutoInstrumentation.StartupHook/StartupHook.cs:103]: https://github.com/dotnet/runtime/blob/main/docs/design/features/host-startup-hook.md#no-dependency-resolution-for-non-app-assemblies kudos to @lachmatt for finding it 😉

**Comment by pjanotti** [comment]
@dszmigielski The fact that we need an explicit load from the `StartupHook` and the fact this introduces a project dependency to the managed instrumentation project makes me think if it wouldn't be better to just share the sources instead of creating a new project. I don't see a major problem with adding a new dll, but, OTOH I don't see any downside to just share the sources between the projects. The sources could go under something like `./src/shared/OpenTelemetry.AutoInstrumentation.Logging/` .

**Review by pjanotti — CHANGES_REQUESTED** [review]
Please, see https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/1983#issuecomment-1387684812

**Comment by dszmigielski** [comment]
>

**Comment by dszmigielski** [comment]
> @dszmigielski The fact that we need an explicit load from the `StartupHook` and the fact this introduces a project dependency to the managed instrumentation project makes me think if it wouldn't be better to just share the sources instead of creating a new project. I don't see a major problem with adding a new dll, but, OTOH I don't see any benefits versus sharing the sources between the 2 projects. The sources could go under something like `./src/shared/OpenTelemetry.AutoInstrumentation.Logging/` .

done, please review again :)

File: src/OpenTelemetry.AutoInstrumentation.StartupHook/Logger.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace OpenTelemetry.AutoInstrumentation.StartupHook;
+
+internal static class Logger
+{
+    private static readonly IOtelLogger Log = OtelLogging.GetLogger();
+
+    internal static void LogInformation(string message)
+    {
+        StartupHookEventSource.Log.Trace(message);
```
Comment (pjanotti) [inline src/OpenTelemetry.AutoInstrumentation.StartupHook/Logger.cs:28]: nit of nitness: just for source consistency: the other `Logger.cs`, combining EventSource with the `OtelLogging`, does it in the reverse order. Use the same order on both.

**Review by pjanotti — COMMENTED** [review]
Thanks for moving to shared sources. A few small things to change before we can merge.

File: src/OpenTelemetry.AutoInstrumentation.AspNetCoreBootstrapper/Logger.cs (hunk 3 above)
Comment (rajkumar-rangaraj) [reply src/OpenTelemetry.AutoInstrumentation.AspNetCoreBootstrapper/Logger.cs:28]: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2017

File: src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs [hunk 12]
```diff
@@ -42,7 +42,7 @@ public void Add(InstrumentationInitializer loader)
 
     private class OnAssemblyLoadInitializer
     {
-        private static readonly ILogger Logger = OtelLogging.GetLogger();
+        private static readonly IOtelLogger OtelLogger = OtelLogging.GetLogger();
```
[the remark this answers is not in the document]
Comment (rajkumar-rangaraj) [reply src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs:45]: It was based on my feedback. I agree, its better to rename in a follow up PR.

**Comment by rajkumar-rangaraj** [comment]
@dszmigielski 
Files are moved into `src/Shared/OpenTelemetry.AutoInstrumentation.Logging/` without a project file. It's best practice to create shared project and place files in it.

File: src/OpenTelemetry.AutoInstrumentation/Properties/AssemblyInfo.NetFramework.cs [hunk 13]
```diff
@@ -18,7 +18,7 @@
 using System.Reflection;
 using System.Runtime.CompilerServices;
 
-[assembly: AssemblyKeyFile("keypair.snk")]
+[assembly: AssemblyKeyFile("../keypair.snk")]
```
[the remark this answers is not in the document]
Comment (dszmigielski) [reply src/OpenTelemetry.AutoInstrumentation/Properties/AssemblyInfo.NetFramework.cs]: > @dszmigielski Files are moved into `src/Shared/OpenTelemetry.AutoInstrumentation.Logging/` without a project file. It's best practice to create shared project and place files in it.
> 
> It is not a blocker and we could create an issue to track it.

I would leave it as is for the moment. It will require build changes, as we would probably not want that to be built, and for sure we would not want the output dll to be included into tracer.

**Comment by dszmigielski** [comment]
> @dszmigielski Files are moved into `src/Shared/OpenTelemetry.AutoInstrumentation.Logging/` without a project file. It's best practice to create shared project and place files in it.
> 
> It is not a blocker and we could create an issue to track it.

I would leave it as is for the moment. It will require build changes, as we would probably not want that to be built, and for sure we would not want the output dll to be included into tracer.

File: src/OpenTelemetry.AutoInstrumentation.Logging/CustomLogger.cs (hunk 5 above)
Comment (pellared) [reply src/OpenTelemetry.AutoInstrumentation.Logging/CustomLogger.cs]: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2022

**Comment by dszmigielski** [comment]
@rajkumar-rangaraj after consulting with @pellared I have moved logger source files back to `OpenTelemetry.AutoInstrumentation` and linked them to other project. I hope that this solution works for you.

**Comment by pellared** [comment]
> @rajkumar-rangaraj after consulting with @pellared I have moved logger source files back to `OpenTelemetry.AutoInstrumentation` and linked them to other project. I hope that this solution works for you.

I just want to mention that this approach is used in https://github.com/open-telemetry/opentelemetry-dotnet

File: src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (pellared) [reply src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs:45]: Maybe as part of https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2022 ?

File: src/OpenTelemetry.AutoInstrumentation/CallTarget/Handlers/IntegrationMapper.cs [hunk 14]
```diff
@@ -28,7 +28,7 @@ internal class IntegrationMapper
     private const string EndMethodName = "OnMethodEnd";
     private const string EndAsyncMethodName = "OnAsyncMethodEnd";
 
-    private static readonly ILogger Log = OtelLogging.GetLogger();
+    private static readonly IOtelLogger Log = OtelLogging.GetLogger();
```
[the remark this answers is not in the document]
Comment (pellared) [reply src/OpenTelemetry.AutoInstrumentation/CallTarget/Handlers/IntegrationMapper.cs:31]: Added as comment here: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2022

**Comment by rajkumar-rangaraj** [comment]
> uld probably not want that to be built, and for sure we would not want the output dll to be included into tracer.

I was speaking here about `.shproj` and not `.csproj`. I have been following this PR since beginning and completely agree with @pjanotti feedback, we should not introduce another new library.

New proposed approach looks good to me.

**Review by rajkumar-rangaraj — APPROVED** [review]

**Review by pjanotti — APPROVED** [review]
Thanks @dszmigielski
